src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -2576,6 +2593,39 @@ static void start_retriable_subchannel_batches(void* arg, grpc_error* ignored) {
 // LB pick
 //
 
+// The callback to intercept trailing metadata if retries is not enabled
+static void recv_trailing_metadata_ready_for_lb(void* arg, grpc_error* error) {
+  subchannel_batch_data* batch_data = static_cast<subchannel_batch_data*>(arg);
+  grpc_call_element* elem = batch_data->elem;
+  call_data* calld = static_cast<call_data*>(elem->call_data);
+  if (calld->pick.recv_trailing_metadata != nullptr) {
+    *calld->pick.recv_trailing_metadata =
+        batch_data->batch.payload->recv_trailing_metadata
+            .recv_trailing_metadata;
+  }
+  GRPC_CLOSURE_SCHED(
+      calld->pick.recv_trailing_metadata_ready,
+      GRPC_ERROR_REF(error));
+  GRPC_CLOSURE_RUN(
+      calld->original_recv_trailing_metadata_ready,
+      GRPC_ERROR_REF(error));
+}
+
+// Installs a interceptor to inform the lb of the trailing metadata, if needed
+static void maybe_intercept_trailing_metadata_for_lb(
+    void* arg, grpc_transport_stream_op_batch* batch) {
+  subchannel_batch_data* batch_data = static_cast<subchannel_batch_data*>(arg);
+  grpc_call_element* elem = batch_data->elem;
+  call_data* calld = static_cast<call_data*>(elem->call_data);
+  calld->original_recv_trailing_metadata_ready =
+      batch->payload->recv_trailing_metadata.recv_trailing_metadata_ready;
+  GRPC_CLOSURE_INIT(&calld->recv_trailing_metadata_ready_for_lb,
+                    recv_trailing_metadata_ready_for_lb, elem,
+                    grpc_schedule_on_exec_ctx);
```

Comment:
What's the difference between grpc_schedule_on_exec_ctx vs grpc_combiner_scheduler(chand->combiner) ?

Reply:
Good question.  The answer requires some background and a bit of history.

Every thread in C-core has an [`ExecCtx`](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/exec_ctx.h#L88) associated with it.  The `ExecCtx` is a local variable in all functions that enter the C-core API (primarily, everything defined in [`grpc.h`](https://github.com/grpc/grpc/blob/master/include/grpc/grpc.h)), which means that it gets created when a thread enters C-core and destroyed before that thread leaves C-core.  The `ExecCtx` is stored as a thread-local variable and can be accessed by any function within core.

The `ExecCtx` does a number of things these days, but its original purpose was to provide a general mechanism to avoid deadlocks.  The idea was that whenever code in core needed to invoke a closure, instead of doing it directly and risking a deadlock due to any locks being held when invoking the closure, the code would instead add the closure to the `ExecCtx`, which would invoke all of the closures in its list before it was destroyed (i.e., before the thread returns out of C-core).  This is where we got the difference between `GRPC_CLOSURE_RUN()` and `GRPC_CLOSURE_SCHED()` -- the former runs immediately, and the latter stores the closure to be invoked as soon as we're sure no locks are being held.

Next, we introduced the concept of a [combiner](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/combiner.h), which is a synchronization primative that allows us attempt to get a lock and then if the attempt fails, queue the work we would have done under the lock to be done later instead of blocking a thread waiting for the lock (for more info, see the [design sketch](https://github.com/grpc/grpc/blob/master/doc/combiner-explainer.md); note that that doc is being moved in #16959).  Since closures are the main mechanism we have for doing work later, it made sense to provide this new interface via the closure mechanism, which is what led to the introduction of the closure scheduler.

Every closure has a [closure scheduler](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/closure.h#L53) associated with it.  The scheduler provides the implementations of the `run()` and `sched()` methods that are invoked when we use `GRPC_CLOSURE_RUN()` and `GRPC_CLOSURE_SCHED()`, which allows each scheduler to override the behavior of those two methods.

The `grpc_schedule_on_exec_ctx` scheduler is the original behavior of `GRPC_CLOSURE_RUN()` and `GRPC_CLOSURE_SCHED()` -- the former invokes the closure immediately, and the latter saves it on the `ExecCtx` to be invoked before returning out of C-core.

The `grpc_combiner_scheduler()` scheduler is a scheduler that uses a combiner.  `GRPC_CLOSURE_SCHED()` will attempt to grab the combiner lock.  If successful, it will invoke the closure in the current thread; otherwise, it will queue the closure on the combiner and invokie it in whatever thread is currently holding the combiner.  `GRPC_CLOSURE_RUN()` will invoke the closure immediately -- but note that this is intended to be used when already holding the combiner and when invoking a closure that you know is also using the same combiner; Bad Things(tm) happen when either of those two conditions is not met.

In this case, we are currently using the same combiner for everything in the client channel, including all resolver and LB policy work.  We do plan to change that as part of our work to decouple the control plane and data plane synchronization, but that's the way it is for now.  So for now, we should use `grpc_combiner_scheduler(chand->combiner)` here.

Please let me know if you have any questions about any of this.

Reply:
> GRPC_CLOSURE_RUN() will invoke the closure immediately -- but note that this is intended to be used when already holding the combiner and when invoking a closure that you know is also using the same combiner; Bad Things(tm) happen when either of those two conditions is not met.
I see the assers you are referring to now: https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/combiner.cc#L357


> So for now, we should use grpc_combiner_scheduler(chand->combiner) here.
So in effect, in the end state we will split chand->combiner into two (control vs data plane synchronizations)?  And does this mean using grpc_schedule_on_exec_ctx be just wrong then, because it is the thread local closure scheduler and fails to synchronize on the channel object?

Reply:
> > GRPC_CLOSURE_RUN() will invoke the closure immediately -- but note that this is intended to be used when already holding the combiner and when invoking a closure that you know is also using the same combiner; Bad Things(tm) happen when either of those two conditions is not met.
> > 
> I see the assers you are referring to now: https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/combiner.cc#L357

It looks like that assertion would catch a case where we were accidentally using `GRPC_CLOSURE_RUN()` for a closure running on a *different* combiner than the one we're currently executing in.  However, it would not catch the case where we use `GRPC_CLOSURE_RUN()` on a closure that's running on the `ExecCtx` scheduler (or some other scheduler) while running within a combiner; in that case, there's no way for the scheduler to catch it, beacuse the other schedulers' implementations of `GRPC_CLOSURE_RUN()` don't have any knowledge of the combiner scheduler.  If we accidentally did that anywhere, the result would be that we would wind up increasing lock contention by holding the combiner while running code that doesn't need to be run in the combiner.

> > So for now, we should use grpc_combiner_scheduler(chand->combiner) here.
> > 
> So in effect, in the end state we will split chand->combiner into two (control vs data plane synchronizations)?

The main goal is to split the data plane and control plane so that they don't interfere with each other.  However, it is likely that we won't want a single combiner for the control plane work; instead, we will probably want each resolver and each LB policy to provide their own synchronization for the control plane work, so that (a) their respective APIs are simplified by removing the need to coordinate use of the combiner and (b) each object can do its own control plane work in parallel.  But we can discuss this further when we start actively working on this.

> And does this mean using grpc_schedule_on_exec_ctx be just wrong then, because it is the thread local closure scheduler and fails to synchronize on the channel object?

First, note that the `ExecCtx` scheduler is not strictly a thread-local scheduler; there are cases where a closure scheduled on the `ExecCtx` will actually not be executed right away but will instead wait for the next call into core.

Second, as mentioned above, the main reason it's wrong to use the `ExecCtx` scheduler here is that it would execute code that doesn't need the combiner while still holding the combiner.